File: mobile/android/protocol/src/main/kotlin/ai/deepseek/harness/mobile/conversation/Fold.kt
package ai.deepseek.harness.mobile.conversation

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

data class ImagePart(val mediaType: String, val data: String)

data class Bubble(
    val id: String,
    val role: String,
    val text: String,
    val card: String? = null,
    val images: List<ImagePart> = emptyList(),
)

object Fold {
    fun foldEvents(entries: List<JsonElement>): List<Bubble> {
        val rows = mutableListOf<Bubble>()
        var assistant: Bubble? = null
        fun flushAssistant() {
            assistant?.let { rows.add(it) }
            assistant = null
        }
        for (entry in entries) {
            val obj = entry as? JsonObject ?: continue
            val event = obj["event"] as? JsonObject ?: obj
            val type = event["type"]?.jsonPrimitive?.contentOrNull ?: continue
            val seq = event["seq"]?.jsonPrimitive?.contentOrNull ?: ""
            val data = event["data"] as? JsonObject
            when (type) {
                "user/message" -> {
                    val kind = data?.get("source")?.jsonObject?.get("kind")?.jsonPrimitive?.contentOrNull
                    if (kind != "user") continue
                    flushAssistant()
                    val id = data["id"]?.jsonPrimitive?.contentOrNull ?: seq
                    rows.add(
                        Bubble(
                            id = id,
                            role = "user",
                            text = textFromBlocks(data["content"]),
                            images = imagesFromBlocks(data["content"]),
                        ),
                    )
                }
                "assistant/chunk" -> {
                    if (assistant == null) {
                        assistant = Bubble(id = "assistant-$seq", role = "assistant", text = "")
                    }
                    assistant = assistant!!.copy(text = assistant!!.text + chunkText(data))
                }
                "assistant/message" -> {
                    val content = data?.get("message")?.jsonObject?.get("content")
                    assistant = Bubble(
                        id = "assistant-$seq",
                        role = "assistant",
                        text = textFromBlocks(content),
                    )
                    flushAssistant()
                }
                "tool/call" -> {
                    flushAssistant()
                    val card = obj["view"]?.jsonObject?.get("view")?.jsonObject?.get("card")
                        ?.jsonPrimitive?.contentOrNull
                        ?: data?.get("name")?.jsonPrimitive?.contentOrNull
                        ?: "tool"
                    rows.add(
                        Bubble(
                            id = data?.get("callId")?.jsonPrimitive?.contentOrNull ?: seq,
                            role = "tool",
                            text = data?.get("name")?.jsonPrimitive?.contentOrNull.orEmpty(),
                            card = card,
                        ),
                    )
                }
            }
        }
        flushAssistant()
        return rows
    }

    private fun textFromBlocks(blocks: JsonElement?): String {
        val array = blocks as? JsonArray ?: return ""
        return array.joinToString("") { block ->
            val obj = block as? JsonObject ?: return@joinToString ""
            obj["text"]?.jsonPrimitive?.contentOrNull ?: ""
        }
    }

    private fun imagesFromBlocks(blocks: JsonElement?): List<ImagePart> {
        val array = blocks as? JsonArray ?: return emptyList()
        return array.mapNotNull { block ->
            val obj = block as? JsonObject ?: return@mapNotNull null
            if (obj["type"]?.jsonPrimitive?.contentOrNull != "image") return@mapNotNull null
            val media = obj["mediaType"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
            val data = obj["data"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
            ImagePart(mediaType = media, data = data)
        }
    }

    private fun chunkText(data: JsonObject?): String {
        val chunk = data?.get("chunk")
        if (chunk is JsonObject) {
            return chunk["text"]?.jsonPrimitive?.contentOrNull.orEmpty()
        }
        chunk?.jsonPrimitive?.contentOrNull?.let { return it }
        return data?.get("text")?.jsonPrimitive?.contentOrNull.orEmpty()
    }
}

File: mobile/web/conversation/fold.js
function textFromBlocks(blocks) {
  if (!Array.isArray(blocks)) return '';
  return blocks.map((block) => {
    if (!block || typeof block !== 'object') return '';
    if (typeof block.text === 'string') return block.text;
    return '';
  }).join('');
}

// 镜像 Android Fold.kt imagesFromBlocks：user/message 里的 image block 折叠成气泡图片。
function imagesFromBlocks(blocks) {
  if (!Array.isArray(blocks)) return [];
  const images = [];
  for (const block of blocks) {
    if (!block || typeof block !== 'object' || block.type !== 'image') continue;
    if (typeof block.mediaType !== 'string' || typeof block.data !== 'string') continue;
    images.push({ mediaType: block.mediaType, data: block.data });
  }
  return images;
}

function chunkText(data) {
  const chunk = data?.chunk;
  if (typeof chunk === 'string') return chunk;
  if (chunk && typeof chunk.text === 'string') return chunk.text;
  if (typeof data?.text === 'string') return data.text;
  return '';
}

const TOOL_STATUS_LABELS = {
  running: '运行中',
  completed: '完成',
  failed: '失败',
  canceled: '已取消',
};

// 每类 tool detail 折出一行摘要 + 可展开正文；未知形状保底显示类型名。
function toolDetailView(detail) {
  if (!detail || typeof detail !== 'object') return null;
  if (detail.type === 'shell') {
    return {
      summary: detail.command || '',
      body: typeof detail.output === 'string' ? detail.output : '',
      bodyKind: 'code',
    };
  }
  if (detail.type === 'read' || detail.type === 'write') {
    return {
      summary: detail.filePath || '',
      body: typeof detail.content === 'string' ? detail.content : '',
      bodyKind: 'code',
    };
  }
  if (detail.type === 'edit') {
    return {
      summary: detail.filePath || '',
      body: typeof detail.unifiedDiff === 'string' ? detail.unifiedDiff : '',
      bodyKind: 'code',
    };
  }
  if (detail.type === 'search') {
    const counts = [];
    if (Number.isInteger(detail.numMatches)) counts.push(`${detail.numMatches} 处匹配`);
    if (Number.isInteger(detail.numFiles)) counts.push(`${detail.numFiles} 个文件`);
    return {
      summary: [detail.query, counts.join(' · ')].filter(Boolean).join(' — '),
      body: typeof detail.content === 'string'
        ? detail.content
        : (detail.filePaths || []).join('\n'),
      bodyKind: 'code',
    };
  }
  if (detail.type === 'fetch') {
    return {
      summary: detail.url || '',
      body: typeof detail.result === 'string' ? detail.result : '',
      bodyKind: 'text',
    };
  }
  if (detail.type === 'sub_agent') {
    return {
      summary: detail.description || detail.subAgentType || '子智能体任务',
      body: typeof detail.log === 'string' ? detail.log : '',
      bodyKind: 'text',
      childSessionId: typeof detail.childSessionId === 'string' ? detail.childSessionId : '',
    };
  }
  if (detail.type === 'plan') {
    return { summary: '', body: detail.text || '', bodyKind: 'markdown' };
  }
  if (detail.type === 'plain_text') {
    return { summary: detail.label || '', body: detail.text || '', bodyKind: 'text' };
  }
  if (detail.type === 'worktree_setup') {
    return {
      summary: [detail.branchName, detail.worktreePath].filter(Boolean).join(' · ') || '工作树准备',
      body: typeof detail.log === 'string' ? detail.log : '',
      bodyKind: 'code',
    };
  }
  // unknown detail: keep whatever is stringable visible.
  const body = detail.output ?? detail.input ?? null;
  return {
    summary: typeof detail.type === 'string' ? detail.type : '',
    body: typeof body === 'string' ? body : (body == null ? '' : JSON.stringify(body, null, 2)),
    bodyKind: 'code',
  };
}

function foldTimelineItem(item, entry, fallbackId) {
  if (item.type === 'user_message') {
    return {
      id: String(item.messageId || entry.seqStart || fallbackId),
      role: 'user',
      text: String(item.text || ''),
      images: [],
    };
  }
  if (item.type === 'assistant_message') {
    return {
      id: String(item.messageId || entry.seqStart || fallbackId),
      role: 'assistant',
      text: String(item.text || ''),
    };
  }
  if (item.type === 'reasoning') {
    return {
      id: String(entry.seqStart || fallbackId),
      role: 'reasoning',
      text: String(item.text || ''),
    };
  }
  if (item.type === 'tool_call') {
    return {
      id: String(item.callId || entry.seqStart || fallbackId),
      role: 'tool',
      text: String(item.name || ''),
      card: TOOL_STATUS_LABELS[item.status] || item.status || 'tool',
      status: item.status || '',
      detail: toolDetailView(item.detail),
    };
  }
  if (item.type === 'error') {
    return {
      id: String(entry.seqStart || fallbackId),
      role: 'error',
      text: String(item.message || ''),
    };
  }
  if (item.type === 'todo') {
    return {
      id: String(entry.seqStart || fallbackId),
      role: 'todo',
      items: (Array.isArray(item.items) ? item.items : []).map((todo) => ({
        text: String(todo?.text || ''),
        completed: todo?.completed === true,
      })),
    };
  }
  if (item.type === 'compaction') {
    const status = item.status === 'completed'
      ? '上下文已压缩'
      : item.status === 'failed'
        ? `上下文压缩失败${item.error ? `：${item.error}` : ''}`
        : '正在压缩上下文…';
    return { id: String(entry.seqStart || fallbackId), role: 'meta', text: status };
  }
  if (item.type === 'turn_changes') {
    return {
      id: String(entry.seqStart || fallbackId),
      role: 'changes',
      text: String(item.changeSummary || ''),
      files: (Array.isArray(item.changedFiles) ? item.changedFiles : []).map((file) => ({
        path: String(file?.path || ''),
        additions: Number.isInteger(file?.additions) ? file.additions : null,
        deletions: Number.isInteger(file?.deletions) ? file.deletions : null,
      })),
    };
  }
  if (item.type === 'generative_ui') {
    return {
      id: String(entry.seqStart || fallbackId),
      role: 'meta',
      text: `交互组件「${item.title || item.componentId || 'generative UI'}」需要在电脑端查看`,
    };
  }
  // Unknown wire item types stay visible instead of silently vanishing.
  return {
    id: String(entry.seqStart || fallbackId),
    role: 'meta',
    text: `暂不支持的消息类型：${typeof item.type === 'string' ? item.type : '未知'}`,
  };
}

function foldEvents(entries) {
  const rows = [];
  let assistant = null;
  const flushAssistant = () => {
    if (assistant) {
      rows.push(assistant);
      assistant = null;
    }
  };
  for (const entry of entries || []) {
    const item = entry?.item;
    if (item && typeof item === 'object' && typeof item.type === 'string') {
      flushAssistant();
      rows.push(foldTimelineItem(item, entry, rows.length));
      continue;
    }
    const event = entry?.event || entry;
    if (!event || typeof event.type !== 'string') continue;
    if (event.type === 'user/message') {
      if (event.data?.source?.kind !== 'user') continue;
      flushAssistant();
      rows.push({
        id: String(event.data?.id || event.seq),
        role: 'user',
        text: textFromBlocks(event.data?.content),
        images: imagesFromBlocks(event.data?.content),
      });
      continue;
    }
    if (event.type === 'assistant/chunk') {
      if (!assistant) {
        assistant = { id: `assistant-${event.seq}`, role: 'assistant', text: '' };
      }
      assistant.text += chunkText(event.data);
      continue;
    }
    if (event.type === 'assistant/message') {
      assistant = {
        id: `assistant-${event.seq}`,
        role: 'assistant',
        text: textFromBlocks(event.data?.message?.content),
      };
      flushAssistant();
      continue;
    }
    if (event.type === 'tool/call') {
      flushAssistant();
      rows.push({
        id: String(event.data?.callId || event.seq),
        role: 'tool',
        text: String(event.data?.name || ''),
        card: entry.view?.view?.card || event.data?.name || 'tool',
        detail: null,
      });
    }
  }
  flushAssistant();
  return rows;
}

export { foldEvents, toolDetailView };
